### PR TITLE
Closes #10872: Update tuple implementation and pytest_report_teststatus in docs

### LINKED ISSUE #10872: `pytest_report_teststatus` documentation is contradictory
See [here](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.hookspec.pytest_report_teststatus):

image

The example return value is `"rerun", "R", ("RERUN", {"yellow": True})`, while the return type is documented as 

```py
Tuple[str, str, Union[str, Mapping[str, bool]]]
```

which don't match.

I'm assuming the return type should be

```py
Tuple[str, str, Union[str, Tuple[str, Mapping[str, bool]]]]
```

Or something?

Also some clearer documentation on how the third element of the tuple is sued would be great.

### PR DESCRIPTION
**Pull request by roberto-aldera**
This PR updates the tuple structure in `pytest_runtest_logreport` to use a `typing.NamedTuple`, and fixes the documentation discrepancy outlined in the original issue #10872.

### CODE REVIEW

File: src/_pytest/terminal.py
Comment (RonnyPfannschmidt) [inline src/_pytest/terminal.py:552]: If we add this, let's use a typing.NamedTuple and add expected types + use it in type annotation and internally as well

**Review by RonnyPfannschmidt — COMMENTED** [review]
It may be necessary to define a very painful union for a while

File: src/_pytest/terminal.py
Comment (roberto-aldera) [reply src/_pytest/terminal.py]: Didn't know `typing.NamedTuple` was a thing, thanks 👍 I've had a go at updating it now

File: changelog/10872.improvement.rst
Comment (nicoddemus) [inline changelog/10872.improvement.rst:1]:
```suggestion
Update test log report annotation to named tuple and fixed inconsistency in docs for :hook:`pytest_report_teststatus` hook.
```

File: src/_pytest/terminal.py
Comment (nicoddemus) [inline src/_pytest/terminal.py:549]: Can you please add a docstring here, also describing each member?

Comment (nicoddemus) [reply src/_pytest/terminal.py:549]: I suggest we do not declare this class inside `TerminalReporter`, but make it top-level and also available directly from `pytest.TestStatus`.

Also not sure about the name itself, `TestStatus` feels too generic and similar to `TestReport` (although I see where it is named after)... however no good suggestions come to mind either, bringing this up in case someone has a good suggestion.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @roberto-aldera, left some comments, please take a look.

File: src/_pytest/hookspec.py
Comment (RonnyPfannschmidt) [reply src/_pytest/hookspec.py:809]: for sanity/safety we probably need to keep the old type spec there as part of a union for a while (there should be some type of grace period

File: src/_pytest/terminal.py
Comment (RonnyPfannschmidt) [reply src/_pytest/terminal.py:549]: the name may need a litte bikeshedding as this is bacsically the terminalwriter summary test status

Comment (roberto-aldera) [reply src/_pytest/terminal.py:549]: @nicoddemus, just a question on `reference.rst`, how should I be documenting this new structure in there? I found [this section](https://github.com/pytest-dev/pytest/blob/main/doc/en/reference/reference.rst#reporting-hooks) but am not too sure what to add, something like:
```
.. class:: TestStatus
```

As for the name, I'm all ears, I just went for what seemed a good fit with the `pytest_report_teststatus` hook (as you noticed already). Any suggestions?

File: src/_pytest/hookspec.py
Comment (roberto-aldera) [reply src/_pytest/hookspec.py:809]: Ok, so revert this bit? Or update the docstring? Looks like these are opposite suggestions but I suppose caution is best? Do we want to commit to a timeframe on when the update here could be made if we go ahead?

File: src/_pytest/terminal.py
Comment (roberto-aldera) [inline src/_pytest/terminal.py:124]: This is based on what we have in the docstring for the hook over [here](https://github.com/pytest-dev/pytest/blob/23e343af6018ef3dcf8650e9b3c4a8da0fabe566/src/_pytest/hookspec.py#L813-L824) in case that's helpful information ℹ️

File: src/_pytest/terminal.py
Comment (nicoddemus) [reply src/_pytest/terminal.py]: As for the name, how about `TestShortLogReport`?

File: src/_pytest/hookspec.py
Comment (nicoddemus) [reply src/_pytest/hookspec.py:810]: I think @RonnyPfannschmidt means to change the return value to `TerminalReporter.TestStatus | Tuple[str, str, Union[str, Mapping[str, bool]]]`.

We can probably leave that indefinitely.

File: src/_pytest/terminal.py
Comment (nicoddemus) [inline src/_pytest/terminal.py:122]:
```suggestion
    :ivar category: 
        The class of result, for example ``“passed”``, ``“skipped”``, ``“error”``, or the empty string.

    :ivar letter: 
        The short letter shown as testing progresses, for example ``"."``, ``"s"``, ``"E"``, or the empty string.
        
    :ivar word: 
        Verbose word is shown as testing progresses in verbose mode, for example ``"PASSED"``, ``"SKIPPED"``.
```

File: src/_pytest/hookspec.py
Comment (roberto-aldera) [reply src/_pytest/hookspec.py:811]: Ah that's sensible, I've made that change (but to use `Tuple[str, str, Union[str, Tuple[str, Mapping[str, bool]]]]` instead, I think you meant that? I guess this is why we're fixing it 😆)

Comment (roberto-aldera) [reply src/_pytest/hookspec.py:811]: Ah that's sensible, I've made that change (but to use `Tuple[str, str, Union[str, Tuple[str, Mapping[str, bool]]]]` instead, I think you meant that? I guess this is why we're fixing it 😆)

**Comment by roberto-aldera** [comment]
I'm getting a failure locally when building the docs with `tox -e docs`:
```
WARNING: autodoc: failed to import class 'TestShortLogReport' from module 'pytest'; the following exception was raised:
```
etc etc, assuming that'll either fail here too or I just needed to rebuild something else? By commenting it out in `reference.rst` I can silence it, so that part is not happy.

Another thing I noticed: we refer to `pytest_report_teststatus` in about 4 other places and are still using a plain tuple, I guess I should update those instances too in this change?

**Comment by nicoddemus** [comment]
I think you just need to import that class in https://github.com/pytest-dev/pytest/blob/main/src/pytest/__init__.py and add it to `__all__`.

**Review by nicoddemus — APPROVED** [review]
Thanks @roberto-aldera, LGTM! Appreciate the contribution. :tada: 

We should squash the commits before merging (do not worry, we can do that here directly when clicking on the merge button).

**Comment by roberto-aldera** [comment]
Great, thanks @nicoddemus and @RonnyPfannschmidt for the guidance, that was super helpful! ❤️ 
Looks like the merging is up to you (needs repository write access) so are you ok to squash and then merge?

**Comment by nicoddemus** [comment]
> so are you ok to squash and then merge?

Yes, don't worry we will do at the time of the merge.

Will merge tomorrow unless someone wants more time to give this another look. :+1:

**Comment by nicoddemus** [comment]
Thanks again @roberto-aldera for the contribution!

**Pull request merged**
